### PR TITLE
[#8] Add specifications for the Kafka-based Telemetry API and Event API.

### PR DESCRIPTION
**Pull request by b-abel**
Proposal for Kafka-based Telemetry and Events APIs. Required by [#8 ]. 

Hint:
build and preview API specifications with:

```
# from Hono's base directory
cd site/documentation
hugo server -v
```

Then open http://localhost:1313/hono/docs/api/telemetry/#telemetry-api-for-kafka and http://localhost:1313/hono/docs/api/event/#event-api-for-kafka in browser.

### CODE REVIEW

File: site/documentation/content/api/telemetry/index.md
Comment (sophokles73) [inline site/documentation/content/api/telemetry/index.md:206]: which device ID is that? The authenticated device's or the device's that the data originates from?

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:147]: The `/` is not allowed in Kafka topic names. The only valid non-alphanumerical characters are `.`, `_`, and `-`. Underscore and period could collide in metric systems, so I decided to use the safe character.

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:152]: Yes, that's the common term (see e.g. https://kafka.apache.org/documentation/#brokerconfigs).

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:172]: Yes. The Kafka client distinguishes different types of errors. For those that are potentially transient, retries are made as configured (you can configure the maximum retry count and/or a timeout, and you can disable retries), for those that are final, it throws an exception immediately.

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:193]: Yes.

File: site/documentation/content/api/telemetry/index.md
Comment (b-abel) [reply site/documentation/content/api/telemetry/index.md:139]: I like the idea of enabling readers to easily understand the specification without the need to be Kafka experts beforehand. On the other side, we must then make clear that those short explanations of Kafka basics can not be authoritative nor exhaustive. 
I'm not sure where to put this. Maybe in a separate document and link to it from the API specifications?

File: site/documentation/content/api/telemetry/index.md
Comment (b-abel) [reply site/documentation/content/api/telemetry/index.md:206]: The ID of the device that the data originates from. I'm going to add this.

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:158]: You mean having each configuration property explained in one place and then referencing to this where they are mentioned? I like the idea. But it needs to focus on _why_ and _how_ the property is relevant to Hono, because the Kafka documentation already does a great job in explaining each configuration option: https://kafka.apache.org/documentation/#producerconfigs
We need to be careful here, because, in contrast to the AMQP 1.0 specification, Kafka is not a standard and it is constantly evolving.

File: site/documentation/content/api/telemetry/index.md
Comment (b-abel) [reply site/documentation/content/api/telemetry/index.md:168]: I would rather list recommended (combinations of) configuration options than exhaustively listing all other combinations. 
The table is very verbose while the most important information in it is a just reiteration of the three sentences above. The explanations in the last column may be better moved to the explanation of Kafka basics that we discussed above.

File: site/documentation/content/api/event/index.md
Comment (sophokles73) [reply site/documentation/content/api/event/index.md:147]: Tenant IDs that are based on UUIDs will contain several `-` characters. Will this be a problem, e.g. with distinguishing topic names?

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:196]: Absolutely. 
Hono has no way to recognize the behavior of consumers and therefore there are no consequences possible (like with AMQP where Hono could terminate the link). I think the only thing we _can_ do here is stating that consumers are _expected_ to implement AT LEAST ONCE semantics. Even though a device can't be sure that the event will eventually be consumed by an application (as with publish/subscribe messaging there might be 0, 1, or more consumers).

File: site/documentation/content/api/telemetry/index.md
Comment (sophokles73) [reply site/documentation/content/api/telemetry/index.md:139]: FMPOV this could be embedded in the API document itself. We can then also include a pointer to the official Kafka documentation.

File: site/documentation/content/api/event/index.md
Comment (sophokles73) [reply site/documentation/content/api/event/index.md:196]: Then I guess that is what you should explain here.

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:147]: If we strictly adhere to the schema that a topic starts with the prefix consisting of endpoint and `-` we can always only consider the first `-`. The other two characters are also not prohibited in tenant names, right?

Comment (sophokles73) [reply site/documentation/content/api/event/index.md:147]: no, they are not. But in the default configuration the tenant ID pattern excludes `/` ...
As long as the topic names consist of two segments only, we might get away with considering the first `-` as the delimiter. However, with the C&C discussions going on, I am not sure if we can safely assume that it will always only be two segments ...
It would be safer to come up with another idea, e.g. URI escaping the `-` in path segments or something like that.

Comment (b-abel) [reply site/documentation/content/api/event/index.md:147]: Before having the problem of topic names with more than two elements, we have the even more pressing issue here that tenant IDs may contain characters that are not allowed and prevent the topic to be created in the first place.

URI-encoding does not work because the `%` character is not allowed in topic names. To solve this we could invent our own encoding logic. I am hesitating to do this because backend applications would be required to do this as well, which might be error-prone.

A second option is to restrict tenant names (I would suggest to the characters used in UUIDs). Even if we don't want to change the pattern that validates tenant IDs with a minor version of Hono, it could be documented that using Kafka requires migrating tenant IDs to this restricted character set (plus starting, to create only safe tenant IDs in our device registry code).

A third option would be to not use the tenant ID in the topic name, but a new, safe identifier. This could be stored in the tenant configuration. The drawback is that backend applications could no longer derive it from the tenant ID but need somehow to get the topic name.

Comment (sophokles73) [reply site/documentation/content/api/event/index.md:147]: FMPOV option 2 should be ok as we are not restricting any existing functionality but add new functionality here ...

Comment (b-abel) [reply site/documentation/content/api/event/index.md:147]: I just had a look at the tenant ID pattern. This is exactly the same character set that Kafka allows for topic names: https://github.com/eclipse/hono/blob/41ad9e6f73b8bb1cd1dd6742a19c6817b6aa9d94/core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java#L343

So, we **do not** have the problem of invalid characters in current tenant IDs.

We could continue with the initial approach of static prefixes followed by the tenant ID. From my understanding, even for C&C, there should be no need for topics with more (dynamic) elements. 

Alternatively, if we do not feel safe with this approach, I would deprecate the current pattern and add a new one that does not include the dot ('.'). The dot would then be Hono's the delimiter character for topic names. I would still allow '_' because it is used in "DEFAULT_TENANT" which is widely used, even in external resources like blog articles (and of course '-' because of the UUIDs).

@sophokles73 WDYT?
@calohmn, do you expect topic names for C&C to contain multiple elements that would require to have a delimiter character?

File: site/documentation/content/api/event/index.md
Comment (Alfusainey) [reply site/documentation/content/api/event/index.md:172]: @b-abel how about using the word "raises" or "throws" instead of "receives"? When I read this, it first appears as if the producer is receiving an exception (in the form of a message) from the Kafka cluster.

Comment (Alfusainey) [reply site/documentation/content/api/event/index.md:172]: maybe also replacing "when" with "while" since the error is raised during the sending process. Also, how about replacing "the" with "a" since you are referring to any Kafka producer and not a specific one.. e.g `When *a* Kafka producer *raises* an exception *while* sending an event message to Kafka...`

File: site/documentation/content/api/telemetry/index.md
Comment (Alfusainey) [inline site/documentation/content/api/telemetry/index.md:175]: How about using `either` here? I think when you say `both`, then it means both *AT LEAST ONCE* and *AT MOST ONCE*. 

maybe it is worth to distinguish how this implements AT LEAST ONCE or AT MOST ONCE delivery semantics. Do you mean that it implements `AT LEAST ONCE` when there is at least one alive in-sync replica? And it implements `AT MOST ONCE` when no in-sync replica is  alive?

File: site/documentation/content/api/event/index.md
Comment (calohmn) [reply site/documentation/content/api/event/index.md:147]: Concerning C&C, the envisioned approaches only use topics with a fixed prefix followed by the tenant ID or some other ID. So, for these, no delimiter would be needed.

Comment (b-abel) [reply site/documentation/content/api/event/index.md:147]: My proposal for topic names is: 
`hono.${endpoint}.${tenant_id}`.
1. Namespacing (i.e. starting with `hono`) is a good practice and might come in handy later on (e.g. Kafka supports subscriptions on multiple topics with regular expressions.
2. In contrast to my first proposal, I would not use `-` as a delimiter to make it more readable with UUIDs as tenant IDs. Since the `_` is already contained in the famous "DEFAULT_TENANT" and in the endpoint `command_response", the `.` is my first choice.

@sophokles73 @calohmn WDYT?

**Comment by b-abel** [comment]
I pushed an update that incorporates feedback

File: site/documentation/content/api/telemetry/index.md
Comment (b-abel) [reply site/documentation/content/api/telemetry/index.md]: @Alfusainey Thank you for your feedback. Please have a look at my last change if it resolves this issue.

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md:172]: @Alfusainey You are right, this is poorly formulated. I will have a look at this next week.

File: site/documentation/content/api/event/index.md
Comment (sophokles73) [reply site/documentation/content/api/event/index.md:149]: Based on https://stackoverflow.com/questions/37062904/what-are-apache-kafka-topic-name-limitations it seems that we should use *either* `.` or `_` but not both. Given that we already have the `DEFALUT_TENANT` name, it looks like we are advised against using `.` as well then.

How about using `__` (two underscores) as the delimiter instead?

Comment (b-abel) [reply site/documentation/content/api/event/index.md:149]: > we should use _either_ `.` or `_` but not both

That is right. The reason is that they could collide in metrics. Therefore the Kafka CLI always prints a warning if you create a topic that contains a  `.` or a `_`. In practice, there would be no collision with our naming schema.

> How about using `__` (two underscores) as the delimiter instead?

I like the creative thinking here, but the drawback is that two underscores are hard to read. Depending on the font they are displayed as a single long underscore (obviously we need to explain at least here on Github what it is).

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md]: @Alfusainey Thank you for your great suggestion. I applied it.

Comment (b-abel) [reply site/documentation/content/api/event/index.md]: The current specification of the [Telemetry API](https://www.eclipse.org/hono/docs/api/telemetry/) says:

> When the AMQP Messaging Network fails to settle the transfer of a telemetry message or settles the transfer with any other outcome than `accepted`, the protocol adapter MUST NOT try to re-send such rejected messages but SHOULD indicate the failed transfer to the device if the transport protocol provides means to do so.

For the event API it says: "...but MUST indicate the failed transfer...".

@sophokles73 I am not sure about the reasoning behind this and this should be applied to the Kafka based API. Why must the protocol adapter not implement a retry? And does the difference regarding the obligation to inform the device only arise from the fact that this is not possible with QoS 0?

**Comment by sophokles73** [comment]
@b-abel The more often I read through the spec the more I think that I would prefer to have the Kafka based API in separate documents. This makes them easier to find in the documentation and IMHO it also underlines the fact that the two APIs are different and do not share much more than the overall semantics of the message payload. WDYT?

**Comment by b-abel** [comment]
> @b-abel The more often I read through the spec the more I think that I would prefer to have the Kafka based API in separate documents. This makes them easier to find in the documentation and IMHO it also underlines the fact that the two APIs are different and do not share much more than the overall semantics of the message payload. WDYT?

@sophokles73 I feel exactly the same!

**Comment by b-abel** [comment]
I moved the Kafka based APIs to separate documents. 
I found no better names than that (e.g. "Telemetry API for Kafka"). Maybe someone else has a better idea...
The last commit moves the APIs to a new submenu. We do not have submenus yet in the documentation and I'm not sure if it's a good idea. Feedback is welcome.

**Comment by b-abel** [comment]
I just added some changes to highlight that a protocol adapter must not use two Kafka producers for telemetry (for each QoS level). Otherwise, messages from one device with different QoS levels might be out of order. 

I will update #2216 accordingly.

File: site/documentation/content/api/kafka/telemetry-kafka/index.md
Comment (sophokles73) [inline site/documentation/content/api/kafka/telemetry-kafka/index.md:152]: it might be helpful to also explain the reason for this requirement.

File: site/documentation/content/api/event/index.md
Comment (b-abel) [reply site/documentation/content/api/event/index.md]: I think it would be good if we could come to a final decision regarding the structure of topic names soon. FMPOV, we can't complete the API specification or implementation before that. 

My proposal would be the following: 
1. we start with using topic names with a static prefix and a dynamic second element. Then we do not need a reserved delimiter.
2. we separate the two elements (and possible sub-elements within the prefix) with a defined _separator_, which is only for human readability and not used as a _technical delimiter_ for parsing the name.
3. we take the '.' as the _separator_. Although we then have '.' and '_' in topic names, which in principle could lead to name conflicts in Kafka metrics, no conflicts will occur with our structure of the names.
4. if we realize at a later time that this naming scheme is no longer sufficient for us, we can still define a technical delimiter. For the TenantIds, '.' is a good choice for the reasons mentioned above, so that we don't have to change the character again. For other dynamic elements, we can't tell which characters they contain without knowing them anyway.
5. the static prefixes are additionally namespaced with `hono`.

The result will look like this:

`hono.${endpoint}.${tenant_id}`. Or, more concrete: `hono.event.${tenant_id}`, `hono.telemetry.${tenant_id}`, `hono.command.${tenant_id}`, `hono.command_response.${tenant_id}`.

WDYT?

Comment (sophokles73) [reply site/documentation/content/api/event/index.md]: Looks ok FMPOV. What is the proposed pattern for tenant identifiers then?

Comment (b-abel) [reply site/documentation/content/api/event/index.md]: > Looks ok FMPOV. What is the proposed pattern for tenant identifiers then?

I think we could leave it as it is for now. If we consider everything after the fixed prefix as identifiers for the client, it doesn't matter what it contains. But if you feel more comfortable if we now deprecate '.' in the tenant identifiers, that's fine with me too.

File: site/documentation/content/api/telemetry/index.md
Comment (sophokles73) [reply site/documentation/content/api/telemetry/index.md]: No, I agree with the approach. In the end, that is the whole point of defining the *fixed prefix*, right?

Comment (b-abel) [reply site/documentation/content/api/telemetry/index.md]: Yes.

**Comment by b-abel** [comment]
I changed the topic names in the spec like this:

> `hono.${endpoint}.${tenant_id}`. Or, more concrete: `hono.event.${tenant_id}`, `hono.telemetry.${tenant_id}`, `hono.command.${tenant_id}`, `hono.command_response.${tenant_id}`.

**Comment by sophokles73** [comment]
@b-abel IMHO we can merge this PR once you have rebased ...

**Comment by b-abel** [comment]
I added a small change in the last commit. This removes the submenu for the Kafka-based API specifications. Instead, the API specifications are now displayed again at the same level in the menu as all others. We don't use nested submenus anywhere else either and I'm afraid that the display might not be optimal, at least on small screens.
Besides that, it is ready to be merged FMPOV.

**Review by sophokles73 — APPROVED** [review]
LGTM

**Comment by sophokles73** [comment]
@calohmn any additional comments?

**Review by calohmn — APPROVED** [review]
No, LGTM.

**Pull request merged**
